packages/babel-preset-env/src/normalize-options.js

```diff
@@ -165,6 +176,71 @@ export const validateUseBuiltInsOption = (
   return builtInsOpt;
 };
 
+// Remove spec/loose options in next major version
+export const validateTransformModeOption = ({
+  transformMode,
+  spec: specOption,
+  loose: looseOption,
+}: {
+  transformMode?: TransformMode,
+  spec?: boolean,
+  loose?: boolean,
+}) => {
+  const specIsUndefined = typeof specOption === "undefined";
+  const looseIsUndefined = typeof looseOption === "undefined";
+  const transformModeIsUndefined = typeof transformMode === "undefined";
+
+  if (!specIsUndefined) {
+    console.log("");
+    console.log("The top-level `spec` option has been deprecated.");
```

Comment:
IMHO we shouldn't deprecate spec/loose - they map nicely to options accepted by plugins, if we introduce in the future more things that `transformMode` would account for it would still be useful to use spec/loose appropriately (maybe in combination with `transformMode`, maybe not - not sure if it would be possible to use them together at this point)

that's my opinion though, `transformMode` seems like something targeting newcomers, atm hiding spec/loose under more descriptive option, but for more experienced users loose/spec options are good because we can easily relate them to plugins' options

Reply:
Yeah I think it would be rather weird if we left the plugin options to use spec/loose still but the preset uses something different?

Reply:
Yeah, that's what I'm thinking. I see a value in `transformMode` though as some kind of options preset for the future use, although not sure how many more granular (than loose/spec) options we will truly have.

Also there is a question that would have to be IMHO answered here - what should `transformMode === 'performance'` do [here](https://github.com/babel/babel/blob/19a170529350216989384330988560b14eed6ae8/packages/babel-plugin-transform-for-of/src/index.js#L11)? How one can use preset-env with `assumeArray` nowadays? I suspect this might not be possible, but aint sure

Reply:
I think it might be helpful to frame the discussion in terms of the two overarching types of preset users:

**1. Newcomers, general usage**

I think the beauty of the env preset is it is a unifying abstraction that makes the Babel ecosystem extremely easy to use (by providing a declarative, high-level interface that corresponds closely to user intent). So just as inclusion/exclusion of individual plugins is abstracted over with the `targets` option, so should configuration of plugins. I think there are some issues with `spec`/`loose` in that regard:
- They only "loosely" correspond to user intent. `loose` is rather vague and doesn't imply the the actual desired behavior, which is performance.
- It's unclear what happens when `{loose: true, spec: true}` is set. Without knowledge of the underlying plugins (which is very possible given the nature of this preset), this is quite confusing.
- It's unclear what happens when `{loose: false, spec: false}` is set. Again, understanding what this does requires familiarity of conventions surrounding plugin implementation and configuration. Babel experts of course know that there's some default which strikes as balance between perf and spec compliance, but this is totally implicit (in terms of the config `{loose: false, spec: false}`). Instead, I think an explicit option (e.g. `{transformMode: "normal"}` or `{transformMode: "balanced"}` would be much more approachable and intuitive.

I think exposing `spec` and `loose` as the primary means of configuration leaks what should be considered implementation details at this level of abstraction. I think generally, people don't want have to care about configuration of individual plugins. Which brings me to the next use case:

**2. Power users**

I think it is dubious whether top-level `spec`/`loose` options are even sufficient for power users. I think in practice, usually there's one or more specific transformations in specific plugins that a power user wants to configure. For this, top-level `spec`/`loose` are far too broad, instead a more precise configuration mechanism is needed. 

This is why in https://github.com/babel/babel/issues/8195 I propose `options.configOverrides`, which allows for direct configuration of individual plugins. I think `assumeArray` is the perfect use case for this.

Reply:
Ok I generally agree with that, although you think that we shouldn't change the name of the options for individual plugins?

I never really thought the top level options would be sufficient for power users either. We tend to want to include multiple options as it increases in complexity, with the possibility one might want an option that includes them all under an umbrella "loose" mode. This was a suggestion but not sure how it plays out.

Although it's out of scope but related, I'm curious on your thoughts for https://github.com/babel/babel/issues/7770 (and my PR https://github.com/babel/babel/pull/8293). One of the reasons for removing them is the fact we have to pass the options down all the presets (not including updating them all the time, making breaking changes, and people opt-ing into too much or not knowing what's in it). If we remove the yearly/stage presets, preset-env will be the main way people interact with Babel. Should we have a preset for any proposals, (basically decided no now)?

Reply:
I was hesitant to also change the configuration for each plugin in this PR (simply because it would increase the scope of this PR), but I agree with you that it would be better to change the plugin config, if not simply to colocate/encapsulate the "transform mode" behavior/logic within each plugin rather than in the env preset. I think it would be probably best if each plugin decides on its own what the reasonable behavior is for "normal", "performance", and "spec" modes (used by the env preset), but have more granular options that can be turned on/off (when configuring directly). I can update this PR to update the plugin config as well.

Personally, I'm definitely on board with #7770 and agree that the env preset should probably supplant the use of yearly/stage presets. I think not including proposals (at least initially) probably makes sense to me 👍